File: ios/Classes/BeeDynamicLauncherPlugin.swift
import Flutter
import UIKit

public class BeeDynamicLauncherPlugin: NSObject, FlutterPlugin {
  private static let channelName = "dev.bee.bee_dynamic_launcher/launcher"
  private var variantIds: [String] = []
  private var primaryVariantId: String = ""

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger())
    let instance = BeeDynamicLauncherPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      guard let args = call.arguments as? [String: Any],
            let ids = args["ids"] as? [String],
            let primary = args["primaryVariantId"] as? String else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "ids and primaryVariantId required",
            details: nil))
        return
      }
      if ids.isEmpty || primary.isEmpty {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "ids and primaryVariantId required",
            details: nil))
        return
      }
      if !ids.contains(primary) {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "primaryVariantId must be in ids",
            details: nil))
        return
      }
      variantIds = ids
      primaryVariantId = primary
      result(nil)
    case "getAvailableVariants":
      result(variantIds)
    case "getCurrentVariant":
      if variantIds.isEmpty {
        result(nil)
        return
      }
      if UIApplication.shared.supportsAlternateIcons {
        let name = UIApplication.shared.alternateIconName
        result(variantId(fromAlternateIconName: name))
      } else {
        result(primaryVariantId)
      }
    case "applyVariant":
      guard let id = call.arguments as? String else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "variant id required",
            details: nil))
        return
      }
      if variantIds.isEmpty {
        result(
          FlutterError(
            code: "NOT_INITIALIZED",
            message: "call initialize first",
            details: nil))
        return
      }
      guard variantIds.contains(id) else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENT",
            message: "unknown variant id",
            details: nil))
        return
      }
      let iconName = iosAlternateAssetName(for: id)
      UIApplication.shared.setAlternateIconName(iconName) { error in
        if let error = error {
          result(
            FlutterError(
              code: "APPLY_FAILED",
              message: error.localizedDescription,
              details: nil))
        } else {
          result(nil)
        }
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func iosAlternateAssetName(for id: String) -> String? {
    if id == primaryVariantId {
      return nil
    }
    let suffix = pascalCaseForLauncherId(id)
    return "AppIcon\(suffix)"
  }

  private func pascalCaseForLauncherId(_ id: String) -> String {
    if id.isEmpty {
      return id
    }
    return id.split(separator: "_").map { segment in
      let s = String(segment)
      guard let first = s.first else { return "" }
      return String(first).uppercased() + s.dropFirst().lowercased()
    }.joined()
  }

  private func variantId(fromAlternateIconName name: String?) -> String {
    guard let name = name else {
      return primaryVariantId
    }
    for id in variantIds {
      if let alt = iosAlternateAssetName(for: id), alt == name {
        return id
      }
    }
    return primaryVariantId
  }
}

File: android/src/main/kotlin/dev/bee/bee_dynamic_launcher/BeeDynamicLauncherPlugin.kt
package dev.bee.bee_dynamic_launcher

import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.Locale

class BeeDynamicLauncherPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var variantIds: List<String> = emptyList()
    private var primaryVariantId: String = ""

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                val args = call.arguments as? Map<*, *>
                val ids = (args?.get("ids") as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
                val primary = args?.get("primaryVariantId") as? String ?: ""
                if (ids.isEmpty() || primary.isEmpty()) {
                    result.error("INVALID_ARGUMENT", "ids and primaryVariantId required", null)
                    return
                }
                if (!ids.contains(primary)) {
                    result.error("INVALID_ARGUMENT", "primaryVariantId must be in ids", null)
                    return
                }
                variantIds = ids
                primaryVariantId = primary
                result.success(null)
            }
            "getAvailableVariants" -> result.success(variantIds)
            "getCurrentVariant" -> {
                if (variantIds.isEmpty()) {
                    result.success(null)
                    return
                }
                result.success(getCurrentVariant())
            }
            "applyVariant" -> {
                val id = call.arguments as? String
                if (variantIds.isEmpty()) {
                    result.error("NOT_INITIALIZED", "call initialize first", null)
                    return
                }
                if (id.isNullOrEmpty() || id !in variantIds) {
                    result.error("INVALID_ARGUMENT", "unknown variant id", null)
                    return
                }
                try {
                    applyVariant(id)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("APPLY_FAILED", e.message, null)
                }
            }
            else -> result.notImplemented()
        }
    }

    private fun aliasSimpleClassName(id: String): String {
        if (id.isEmpty()) {
            return id
        }
        return id.split('_').joinToString("") { part ->
            if (part.isEmpty()) {
                return@joinToString ""
            }
            val first = part[0]
            val rest = part.substring(1)
            first.titlecase(Locale.getDefault()) + rest.lowercase(Locale.getDefault())
        }
    }

    private fun componentForVariant(id: String): ComponentName =
        ComponentName(context.packageName, "${context.packageName}.Launcher${aliasSimpleClassName(id)}")

    private fun getCurrentVariant(): String {
        val pm = context.packageManager
        for (id in variantIds) {
            val c = componentForVariant(id)
            if (pm.getComponentEnabledSetting(c) == PackageManager.COMPONENT_ENABLED_STATE_ENABLED) {
                return id
            }
        }
        return variantIds.firstOrNull() ?: primaryVariantId
    }

    private fun applyVariant(id: String) {
        val pm = context.packageManager
        for (vid in variantIds) {
            val c = componentForVariant(vid)
            val enabled = vid == id
            pm.setComponentEnabledSetting(
                c,
                if (enabled) {
                    PackageManager.COMPONENT_ENABLED_STATE_ENABLED
                } else {
                    PackageManager.COMPONENT_ENABLED_STATE_DISABLED
                },
                PackageManager.DONT_KILL_APP,
            )
        }
    }

    companion object {
        private const val CHANNEL = "dev.bee.bee_dynamic_launcher/launcher"
    }
}
